### PR TITLE
Distributed embedding layer

### PR DESCRIPTION
**Pull request by timmoon10**
### Distributed embeddings

The `dist_embedding_layer` has similar functionality as the `embedding_layer`, but distributes its embedding weights between processes. Inter-process communication is one-sided and uses OpenSHMEM (on CPU) or NVSHMEM (on GPU). Forward prop involves "getting" embeddings and backprop involves "putting" gradients. Synchronization is  performed with "quiets" and non-blocking barriers (implemented as non-blocking allreduces on a single `float`).

### Sparse SGD

To take advantage of sparse gradients, the distributed embedding layer provides the option to bypass the optimizer (which currently only supports dense gradients) and perform sparse SGD directly on the embedding weights. If enabled, SGD occurs during the layers "update" phase (i.e. in the virtual `update_compute` function). Otherwise, the layer converts sparse gradients to a dense tensor and passes it into the usual optimizer.

This approach is hacky and I would be receptive to ideas on making it cleaner. Ideally, the optimizer infrastructure would support both dense and sparse gradients. However, things are made tricky by the fact that embedding gradients have structure (entire matrix columns are non-zero) that is not quite amenable to the standard sparse matrix formats (CSR, CSC, COO). In addition, the distributed embedding layer synchronizes with non-blocking barriers to hide load imbalances, which should be important when running with many processes or if the embedding distributions are unbalanced. We'd need to think of a way for sparse optimizers to interact with this synchronization.

</details>

### Performance comparison with data-parallel embeddings

If we added support for sparse gradients to the basic embedding layer, we should expect similar asymptotic scaling behavior between the data-parallel and distributed embedding layers (`O(mini_batch_size / num_procs)` runtime per mini-batch step). The distributed implementation also has a performance penalty from inter-process communication during forward prop. In my node2vec experiments with O(10K) embeddings, the data-parallel implementation outperforms the distributed implementation even without supporting sparsity.

The main benefit of the distributed implementation is that its memory capacity is scalable, which will be necessary for problems where the embeddings don't fit on a single GPU. SGD on distributed embeddings is also scalable while it is not on data-parallel embeddings, although I don't expect SGD to be a bottleneck in practice. In my node2vec experiments with O(1M) embeddings, the distributed implementation shows some scaling and generally blows the data-parallel implementation out of the water (although this is likely due to support for sparsity).

### Building

The CPU implementation requires that LBANN be built with OpenSHMEM. It's hacky, but this can be done in CMake by setting `LBANN_HAS_SHMEM` and modifying `CMAKE_CXX_FLAGS`. This is easiest with OpenMPI (or Spectrum MPI) since they already come with OpenSHMEM implementation, so the linking flags are just `-L/path/to/ompi/lib -loshmem`. I've also had success with Sandia OpenSHMEM. In the future, we should consider implementing a CMake module to make this easier.

The GPU implementation requires building with NVSHMEM, which is supported in the CMake build system. It just requires setting `LBANN_WITH_NVSHMEM` and `NVSHMEM_DIR`.

### Miscellaneous comments

- [Bamboo build](https://lc.llnl.gov/bamboo/browse/LBANN-TIM287-2) has no new failures.
- I've written a Bamboo unit test with metric checking and gradient checking, and it passes when I run on Pascal and Lassen. However, it's disabled for now since our build system doesn't have a nice way to build with SHMEM and NVSHMEM is not yet available on all systems.
- For the full commit history, see https://github.com/timmoon10/lbann/pull/6. This PR removes everything related to the node2vec data reader and the node2vec application, which will be merged in separately.

### CODE REVIEW

**Comment by benson31** [comment]
@samadejacobs Re "dist": I think _some_ differentiation in the API is important, since, by my reading, this layer relies on SHMEM. This makes it significantly different from the MPI universe of the rest of the LBANN codebase. In particular, we won't be requiring SHMEM any time soon (it's a real pain; trust me on that, if you haven't tried it yourself), so this is not as "general-purpose" as other layers. We can debate whether "dist" is the best prefix, but I think _some_ API recognition of this unique use-case is important.

**Comment by timmoon10** [comment]
- I agree "DistEmbedding" layer is not a great name, although I think keeping distinct from the embedding layer is the right choice for now. The external API is similar to the basic embedding layer, but the implementation is more complicated due to SHMEM. It may be more accurate to call the layer "ModelParallelEmbedding", but it doesn't match LBANN's current usage of the term "model-parallelism" (i.e. the data is not in `(MC,MR)` layout).
- The main benefit of this layer is reducing memory pressure from embeddings, which should be helpful both when the vocab size is large or the embedding dim is large. Large embeddings should also have other effects, mainly shifting the communication away from the latency regime toward the bandwidth regime. This would make the performance hit from inter-process communication more painful, but also improve scaling behavior. This would be interesting to look into.
- Copy-pasting these notes into the documentation is a good idea.

File: src/layers/misc/dist_embedding.cu [hunk 1]
```diff
@@ -0,0 +1,732 @@
...
+  }
+  if (m_metadata_buffer != nullptr) {
+    nvshmem_free(m_metadata_buffer);
+  }
+#endif // LBANN_HAS_NVSHMEM
+}
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+void dist_embedding_layer<TensorDataType,Layout,Device>::attach_embeddings_to_shmem_buffer() {
+#ifndef LBANN_HAS_NVSHMEM
```
Comment (benson31) [inline src/layers/misc/dist_embedding.cu:153]: I don't care for this style of cluttering the implementation of nearly every functional part of the code with a preprocessor switch and runtime error. This layer *requires* NVSHMEM or OpenSHMEM. It should not be constructible if you don't have at least one of those, so the runtime errors should never even be reachable. 

I would prefer a more static approach to this -- only compile this class in at all if you have one of those. Then C++ code will never be able to compile code that statically depends on it, and "dynamic code" will never be able to successfully create it because it will not have been registered with the factory.

File: src/layers/misc/dist_embedding.cpp [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+#ifdef LBANN_HAS_SHMEM
+  shmem_free(m_embeddings_buffer);
+  shmem_free(m_workspace_buffer);
+  shmem_free(m_metadata_buffer);
+#endif // LBANN_HAS_SHMEM
+}
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+void dist_embedding_layer<TensorDataType,Layout,Device>::attach_embeddings_to_shmem_buffer() {
+#ifndef LBANN_HAS_SHMEM
```
Comment (benson31) [inline src/layers/misc/dist_embedding.cpp:60]: See comment on corresponding line of `.cu` file.

File: include/lbann/layers/misc/dist_embedding.hpp [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+namespace dist_embedding_layer_impl {
+
+  /** Metadata for an embedding vector from a remote process. */
+  struct vector_metadata {
+    size_t source_rank{0};
+    size_t source_index{0};
+    size_t target_rank{0};
+    size_t target_index{0};
+    bool is_active{false};
+  };
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/layers/misc/dist_embedding.hpp:46]: 1. I want the struct to be available to available to CUDA kernels in an anonymous namespace. The alternative is to put the CUDA kernels in the header file (either as member functions or friends).
2. I like making zero the default value so I can reset a buffer with a memset. It's not a big deal though for the ranks and indices, since they're only read if `is_active` is `true`.

File: include/lbann/layers/misc/dist_embedding.hpp [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+  /** Size of dictionary of embeddings. */
+  size_t m_num_embeddings;
+  /** Size of embedding vectors. */
+  size_t m_embedding_dim;
+
+  /** Perform sparse SGD during backprop.
+   *
+   *  Bypasses optimizer class.
+   */
+  bool m_sparse_sgd;
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/layers/misc/dist_embedding.hpp:120]: For convenience, I like putting all of the user-provided parameters together in the same order they're exposed in the API. It won't result in ideal packing, but putting them after the other class members may be an improvement.

[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/layers/misc/dist_embedding.hpp:120]: This is a fair decision -- I saw the `.proto` so I understand where the ordering came from. In the scope of C++, though, the 1:1 compatibility is less meaningful -- if you're reading C++, you're doing it for a reason and can deal with sort.

All that said, there's likely to be what, 1 of these per model? So we're talking about like 14 wasted bytes per process? If that pushes us to OOM, we've got other issues. And there are far more than 14 bytes being wasted elsewhere we could clean up elsewhere (just as a general comment, not related to this PR).

File: include/lbann/layers/misc/dist_embedding.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/layers/misc/dist_embedding.hpp:46]: 1. Ok, then I would put them inside the struct (they're still impl details of this class), but make them public. We have a bad habit of making things globally available that, at the very least, don't need to be. It makes things look cluttered. Yes, our code is open-source, so anyone can do anything (up to whatever limits are in the license, whatever), but we can at least discourage them from doing bad things.
2. I'm not sure this is going to be faster than anything else (`operator=`); in any case, I care a lot less about the default values if the struct is tucked inside the class.

File: src/base.cpp [hunk 5]
```diff
@@ -104,13 +109,16 @@ world_comm_ptr initialize(int& argc, char**& argv, int seed) {
   init_random(seed);
   init_data_seq_random(seed);
 
-#ifdef LBANN_HAS_NVSHMEM
-  // Initialize NVSHMEM
-  // tym (3/3/20): I get an error when initializing NVSHMEM with
-  // anything other than MPI_COMM_WORLD.
-  // nvshmem::initialize(comm->get_trainer_comm().GetMPIComm()); /// @todo Restore
-  nvshmem::initialize(MPI_COMM_WORLD);
-#endif // LBANN_HAS_NVSHMEM
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply src/base.cpp:113]: I've changed the behavior to only initialize NVSHMEM when it's needed, e.g. the first time `nvshmem_malloc` is called. This approach still initializes with `MPI_COMM_WORLD`: https://github.com/LLNL/lbann/blob/f6271c5b220934933be7205184987bf2bf0c913a/include/lbann/utils/nvshmem.hpp#L52

File: include/lbann/utils/nvshmem.hpp [hunk 6]
```diff
@@ -28,20 +28,88 @@
...
+/** @brief Finalize NVSHMEM library.
+ *
+ *  Does nothing if NVSHMEM has not been initialized or has already
+ *  been finalized. This is _not_ thread-safe.
+ */
 void finalize();
 
+/** @brief Allocate GPU buffer on the NVSHMEM symmetric heap.
+ *
+ *  Initializes NVSHMEM if needed.
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/utils/nvshmem.hpp:63]: Hm, that's a good point. It will add some extra complexity for users though, since they'll need to remember to configure LBANN to initialize NVHSMEM rather than LBANN handling it automatically.

File: src/base.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (benson31) [reply src/base.cpp:113]: Yeah, I saw that after adding the comment. I was replying to commentary that I realize is in the deleted portion.

File: include/lbann/layers/misc/dist_embedding.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/layers/misc/dist_embedding.hpp:120]: Every byte is precious.

File: include/lbann/utils/nvshmem.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/utils/nvshmem.hpp:63]: Can this not be done in `lbann::initialize()` if `LBANN_HAS_NVSHMEM`?

[the remark this answers is not in the document]
Comment (timmoon10) [reply include/lbann/utils/nvshmem.hpp:63]: Do we want to initialize NVSHMEM every time though?

[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/utils/nvshmem.hpp:63]: That's a fair point. On second thought, let's leave this alone and have a discussion offline about cleaning up initialization. There are some other issues tied up in here, like the fact that runs that do use (NV)SHMEM will require more care since mallocs (that dispatch directly to `(nv)shmem_malloc`) are collective on the universe.

File: include/lbann/utils/nvshmem.hpp [hunk 7]
```diff
@@ -28,20 +28,88 @@
...
+// =============================================
+// Implementation
+// =============================================
+
+namespace lbann {
+namespace nvshmem {
+
+template <typename T>
+T* malloc(size_t size) {
+  initialize();
```
[the remark this answers is not in the document]
Comment (benson31) [reply include/lbann/utils/nvshmem.hpp:87]: This can be ignored, pending a larger conversation on initializing these resources.

File: src/utils/nvshmem.cu [hunk 8]
```diff
@@ -26,22 +26,49 @@
...
+
+bool is_active_ = false;
+
+} // namespace <anon>
+
+bool is_active() noexcept {
+  return is_active_;
+}
+
 void initialize(MPI_Comm comm) {
```
[the remark this answers is not in the document]
Comment (benson31) [reply src/utils/nvshmem.cu:43]: Just to clarify, I mean it's an error to reinit after finalize:
```cpp
nvshmem::initialize();
nvshmem::finalize();
nvshmem::initialize();
```
I'm not worried about
```cpp
nvshmem::initialize();
nvshmem::initialize();
nvshmem::initialize();
```
```

**Comment by timmoon10** [comment]
I've updated this PR to accommodate PR #1546. In particular, I'm using the [helper function to get mutable weights values ](https://github.com/LLNL/lbann/blob/develop/src/layers/regularizers/batch_norm_helpers.hpp), which I've moved to `include/lbann/weights/weights_helper.hpp`.
